File: backend/app/schemas/character.py
from pydantic import BaseModel


class EnrichedStartingItem(BaseModel):
    id: int | None = None
    name_en: str
    name_cn: str
    image_url: str | None


class CharacterResponse(BaseModel):
    id: int
    name_en: str
    name_cn: str
    is_tainted: bool
    base_character_id: int | None
    health: str
    damage: str | None
    speed: str | None
    tears: str | None
    starting_items: list | None
    starting_items_enriched: list[EnrichedStartingItem] | None = None
    unlock_method: str | None
    description: str | None
    suitable_items: list | None
    image_url: str | None

    model_config = {"from_attributes": True}


class CharacterListData(BaseModel):
    items: list[CharacterResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/types/character.ts
export interface EnrichedStartingItem {
  id: number | null;
  name_en: string;
  name_cn: string;
  image_url: string | null;
}

export interface Character {
  id: number;
  name_en: string;
  name_cn: string;
  is_tainted: boolean;
  base_character_id: number | null;
  health: string;
  damage: string | null;
  speed: string | null;
  tears: string | null;
  starting_items: number[] | null;
  starting_items_enriched?: EnrichedStartingItem[] | null;
  unlock_method: string | null;
  description: string | null;
  suitable_items: number[] | null;
  image_url: string | null;
}
